File: 1971. Find if Path Exists in Graph/1971. Find if Path Exists in Graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function (n, edges, source, destination) {
  const adjList = new Map();

  for (const [u, v] of edges) {
    if (!adjList.get(u)) {
      adjList.set(u, []);
    }

    if (!adjList.get(v)) {
      adjList.set(v, []);
    }

    adjList.get(u).push(v);
    adjList.get(v).push(u);
  }

  const seen = new Set();
  seen.add(source);
  const stack = [source];

  while (stack.length > 0) {
    const node = stack.pop();

    if (node === destination) {
      return true;
    }

    for (const nei of adjList.get(node) || []) {
      if (!seen.has(nei)) {
        seen.add(nei);
        stack.push(nei);
      }
    }
  }

  return false;
};

File: 1971. Find if Path Exists in Graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if source == destination:
            return True

        adj_list = defaultdict(list)
        for n1, n2 in edges:
            adj_list[n1].append(n2)
            adj_list[n2].append(n1)

        if source not in adj_list or destination not in adj_list:
            return False

        visited = set([source])

        def dfs(node):
            if node == destination:
                return True

            for nei in adj_list[node]:
                if nei not in visited:
                    visited.add(nei)
                    if dfs(nei):
                        return True

            return False

        return dfs(source)
